### PR TITLE
Document that the Eq instances cannot guarantee substitutivity

### PR DESCRIPTION
**Pull request by sjakobi**

### CODE REVIEW

**Comment by sjakobi** [comment]
Improvements on my wording are very welcome!

**Comment by treeowl** [comment]
Has Haddock improved its rendering of instance documentation any? If not, we may want said documentation to just refer to a note (with examples!) elsewhere.

**Comment by sjakobi** [comment]
> Has Haddock improved its rendering of instance documentation any?

Here's a screenshot from the haddocks generated with a pre-release version of ghc-8.8:

image

I agree that examples would help. I'll add some!

**Comment by Shimuuar** [comment]
Explaining what substitutivity is would be very nice too

**Comment by sjakobi** [comment]
Sorry for the delay! This is ready for review now.

**Comment by sjakobi** [comment]
The CI failure should hopefully be resolved by https://github.com/tibbe/unordered-containers/pull/243.

**Comment by sjakobi** [comment]
> The CI failure should hopefully be resolved by #243.

Now it is. :)

File: Data/HashMap/Base.hs [hunk 1]
```diff
@@ -286,6 +286,19 @@ instance Eq k => Eq1 (HashMap k) where
...
+-- >>> data D = A | B deriving (Eq, Show)
+-- >>> instance Hashable D where hashWithSalt salt d = salt
+--
+-- >>> fromList [(A,1), (B,2)] == fromList [(B,2), (A,1)]
+-- True
+--
+-- >>> fromList [(A,1), (B,2)]
+-- fromList [(A,1),(B,2)]
+-- >>> fromList [(B,2), (A,1)]
+-- fromList [(B,2),(A,1)]
```
Comment (treeowl) [inline Data/HashMap/Base.hs]: I don't like using the `Show` instance as an example; the `"fromList"` it adds is confusing. Why not use `toList`?

Comment (sjakobi) [reply Data/HashMap/Base.hs]: Done. I'm also using variables now, since that seemed to make the examples more readable.

**Comment by sjakobi** [comment]
Ping! :)

**Comment by treeowl** [comment]
Could you expand on this to explain just which functions/methods violate substitutivity? These are generally `Foldable`, `Traversable`, and related things.

**Comment by sjakobi** [comment]
> Could you expand on this to explain just which functions/methods violate substitutivity? These are generally `Foldable`, `Traversable`, and related things.

Done. Please check. :)

**Comment by emilypi** [comment]
Seems clear to me with the examples @sjakobi!

**Comment by treeowl** [comment]
I'm heading out to North Carolina right now. Do you think you could check if the functions and instances that violate substitutivity are each documented as such?

**Comment by sjakobi** [comment]
> Do you think you could check if the functions and instances that violate substitutivity are each documented as such?

The `Foldable` and `Traversable` instances, `foldrWithKey`, etc. don't have any documentation in that regard, if that's what you mean. Substitutivity or lack thereof is IMHO primarily a property of the `Eq` instance, so I'd argue that documenting it on the `Eq` instance itself should be sufficient.

**Comment by emilypi** [comment]
Perhaps just mentioning (or linking) the `Eq` docs would be sufficient, since it'd then be searchable. It's very clearly documented as part fo the [laws for `Eq`](http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Eq.html).

**Comment by treeowl** [comment]
Hrmm.... I guess I see `Eq` as a more fundamental thing? It's a very good match to the abstraction. If we were to remove the folds and traversals, then we'd have a perfectly good and intuitive `Eq` situation. Alternatively, we could impose `Ord` on the folds and traversals and recover good behavior by sorting the collision buckets.

**Comment by sjakobi** [comment]
> Perhaps just mentioning (or linking) the `Eq` docs would be sufficient, since it'd then be searchable. It's very clearly documented as a [property of `Eq`](http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Eq.html).

The instance documentation already includes a link to `Eq`.

@treeowl I don't necessarily disagree, but I'm not keen on putting more work into this  PR right now. Would you mind if we simply merge this for now? We can always improve things further in future patches. Also feel free to open issues to track any ideas!

**Comment by emilypi** [comment]
Any progress on this @sjakobi and @treeowl? Documentation can always be improved iteratively. I vote to merge the thing now that it's defined, and call it a day.

**Pull request merged**